doc/configuration/configuration-v1.3.md
# Configuration v1.3

Time to configure flowspy.

First of all you have to copy the dist files to their proper position:

	# cd /srv/flowspy/flowspy
	# cp settings.py.dist settings.py
	# cp urls.py.dist urls.py

Then, you have to edit the settings.py file to correspond to your needs. The settings one has to focus on are:

## Settings.py
Its time to configure `settings.py` in order to connect flowspy with a database, a network device and a broker.

So lets edit settings.py file.

It is strongly advised that you do not change the following to False
values unless, you want to integrate {{ product_name_short }} with you CRM or members
database. This implies that you are able/have the rights to create
database views between the two databases:

    PEER_MANAGED_TABLE = True
    PEER_RANGE_MANAGED_TABLE = True
    PEER_TECHC_MANAGED_TABLE = True

By doing that the corresponding tables as defined in peers/models will
not be created. As noted above, you have to create the views that the
tables will rely on.

### Administrators

	ADMINS: set your admin name and email (assuming that your server can send notifications)

### Secret Key
Please put a random string in `SECRET_KEY` setting.
Make this *unique*, and don't share it with anybody. It's the unique identifier of this instance of the application.

### Allowed hosts
A list of strings representing the host/domain names that this Django site can serve. This is a security measure to prevent an attacker from poisoning caches and password reset emails with links to malicious hosts by submitting requests with a fake HTTP Host header, which is possible even under many seemingly-safe webserver configurations.

For example:

	ALLOWED_HOSTS = ['*']

### Protected subnets
Subnets for which source or destination address will prevent rule creation and notify the `NOTIFY_ADMIN_MAILS`.

	PROTECTED_SUBNETS = ['10.10.0.0/16']


### Database
`DATABASES` should contain the database credentials:

	DATABASES = {
	    'default': {
	        'ENGINE': 'django.db.backends.mysql',
	        'NAME': 'flowspy',
	        'USER': '<db user>',
	        'PASSWORD': '<db password>',
	        'HOST': '<db host>',
	        'PORT': '',
	    }
	}

### Localization
By default Flowspy has translations for English and Greek. In case you want to add
another language, or remove one of the existing, you can change the `LANGUAGES`
variable and follow [django's localization documentation](https://docs.djangoproject.com/en/1.4/topics/i18n/translation/#localization-how-to-create-language-files)

You might want to change `TIME_ZONE` setting too. Here is a [list](http://en.wikipedia.org/wiki/List_of_tz_database_time_zones)


### Cache
Flowspy uses cache in order to be fast. We recomend the usage of memcached, but
any cache backend supported by django should work fine.

	CACHES = {
	    'default': {
	        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
	        'LOCATION': '127.0.0.1:11211',
	    }
	}

### Network device access
We have to inform django about the device we set up earlier.

	NETCONF_DEVICE = "device.example.com"
	NETCONF_USER = "<netconf user>"
	NETCONF_PASS = "<netconf password>"
	NETCONF_PORT = 830


### Beanstalkd
Beanstalk configuration (as a broker for celery)

	BROKER_HOST = "localhost"
	BROKER_PORT = 11300
	POLLS_TUBE = 'polls'
	BROKER_URL = "beanstalk://localhost:11300//"

### Notifications
Outgoing mail address and prefix.

	SERVER_EMAIL = "Example {{ product_name_short }} Service <noreply@example.com>"
	EMAIL_SUBJECT_PREFIX = "[{{ product_name_short }}] "
	NOTIFY_ADMIN_MAILS = ["admin@example.com"]


If you have not installed an outgoing mail server you can always use your own account (either corporate or gmail, hotmail ,etc) by adding the
following lines in settings.py:

    EMAIL_USE_TLS = True #(or False)
    EMAIL_HOST = 'smtp.example.com'
    EMAIL_HOST_USER = 'username'
    EMAIL_HOST_PASSWORD = 'yourpassword'
    EMAIL_PORT = 587 #(outgoing)

### Whois servers
Add two whois servers in order to be able to get all the subnets for an AS.

	PRIMARY_WHOIS = 'whois.example.com'
	ALTERNATE_WHOIS = 'whois.example.net'

### Branding
Fill your company's information in order to show it in flowspy.

	BRANDING = {
	    'name': 'Example.com',
	    'url': 'https://example.com',
	    'footer_iframe': 'https://example.com/iframes/footer/',
	    'facebook': '//facebook.com/example.com',
	    'twitter': '//twitter.com/examplecom',
	    'phone': '800-12-12345',
	    'email': 'helpdesk@example.com',
	    'logo': 'logo.png',
	    'favicon': 'favicon.ico',
	}


### Shibboleth
Flowspy supports shibboleth authentication.

	SHIB_AUTH_ENTITLEMENT = 'urn:mace'
	SHIB_ADMIN_DOMAIN = 'example.com'
	SHIB_LOGOUT_URL = 'https://example.com/Shibboleth.sso/Logout'
	SHIB_USERNAME = ['HTTP_EPPN']
	SHIB_MAIL = ['mail', 'HTTP_MAIL', 'HTTP_SHIB_INETORGPERSON_MAIL']
	SHIB_FIRSTNAME = ['HTTP_SHIB_INETORGPERSON_GIVENNAME']
	SHIB_LASTNAME = ['HTTP_SHIB_PERSON_SURNAME']
	SHIB_ENTITLEMENT = ['HTTP_SHIB_EP_ENTITLEMENT']

### Syncing the database
To create all the tables needed by {{ product_name_short }} we have to run the following commands:

	cd /srv/flowspy
	./manage.py syncdb --noinput
	./manage.py migrate

## Create a superuser
A superuser can be added by using the following command from `/srv/flowspy/`:

	./manage.py createsuperuser


## Propagate the flatpages
Inside the initial\_data/fixtures\_manual.xml file we have placed 4
flatpages (2 for Greek, 2 for English) with Information and Terms of
Service about the service. To import the flatpages, run from root
folder:

    python manage.py loaddata initial_data/fixtures_manual.xml

### Celery
Celery is a distributed task queue, which helps {{ product_name_short }} run some async tasks, like applying a flowspec rule to a router.

`Note` In order to check if celery runs or even debug it, you can run:

	./manage.py celeryd --loglevel=debug


### Testing/Debugging
In order to see what went wrong you can check the following things.

#### Django
You can start the server manually by running:

	./manage.py runserver 127.0.0.1:8081

By doing so, you can serve your application like gunicord does just to test that its starting properly. This command should not be used in production!

Of course you have to stop gunicorn and make sure that port 8081 is free.

#### Gunicorn
Just curl from the server http://localhost:8081

#### Celery
In order to check if celery is working properly one can start celery by typing:

	./manage.py celeryd --loglevel=debug

Again this is for debug purposes.


#### Connectivity to flowspec device
Just try to connect with the credentials you entered in settings.py from the host that will be serving flowspy.


#### General Test
Log in to the admin interface via https://<hostname>/admin. Go to Peer ranges and add a new range (part of/or a complete subnet), eg. 10.20.0.0/19 Go to Peers and add a new peer, eg. id: 1, name: Test, AS: 16503, tag: TEST and move the network you have created from Available to Chosen. From the admin front, go to User, and edit your user. From the bottom of the page, select the TEST peer and save. Last but not least, modify as required the existing (example.com) Site instance (admin home->Sites). You are done. As you are logged-in via the admin, there is no need to go through Shibboleth at this time. Go to https://<hostname>/ and create a new rule. Your rule should be applied on the flowspec capable device after aprox. 10 seconds.

## Footer
Under the templates folder (templates), you can alter the footer.html
file to include your own footer messages, badges, etc.

## Welcome Page
Under the templates folder (templates), you can alter the welcome page -
welcome.html with your own images, carousel, videos, etc.

## Usage

### Web interface
{{ product_name_short }} comes with a web interface, in which one can edit and apply new routes.

### Rest Api
{{ product_name_short }} provides a rest api. It uses token as authentication method.

### Generating Tokens
A user can generate a token for his account on "my profile" page from {{ product_name_short }}'s
UI. Then by using this token in the header of the request he can list, retrieve,
modify and create rules.

### Example Usage
Here are some examples:

#### GET items
- List all the rules your user has created (admin users can see all the rules)

            curl -X GET https://fod.example.com/api/routes/ -H 'Authorization: Token <Your users token>'

- Retrieve a specific rule:

            curl -X GET https://fod.example.com/api/routes/<rule_id>/ -H 'Authorization: Token <Your users token>'

- In order to create or modify a rule you have to use POST/PUT methods.

#### POST/PUT rules
In order to update or create rules you can follow this example:

##### Foreign Keys
In order to create/modify a rule you have to connect the rule with some foreign keys:

###### Ports, Fragmentypes, protocols, thenactions
When creating a rule, one can specify:

- source port
- destination port
- port (if source = destination)

That can be done by getting the url of the desired port instance from `/api/ports/<port_id>/`

Same with Fragmentypes in `/api/fragmenttypes/<fragmenttype_id>/`, protocols in `/api/matchprotocol/<protocol_id>/` and then actions in `/api/thenactions/<action_id>/`.

Since we have the urls we want to connect with the rule we want to create, we can make a POST request like the following:


      curl -X POST -H 'Authorization: Token <Your users token>' -F "name=Example" -F "comments=Description" -F "source=0.0.0.0/0" -F "sourceport=https://fod.example.com/api/ports/7/" -F "destination=203.0.113.12" https://fod.example.com/api/routes/

And here is a PUT request example:

      curl -X PUT -F "name=Example" -F "comments=Description" -F "source=0.0.0.0/0" -F "sourceport=https://fod.example.com/api/ports/7/" -F "destination=83.212.9.93" https://fod.example.com/api/routes/12/ -H  'Authorization: Token <Your users token>'




ChangeLog
=======
1.8.0 RELEASE (2023/05)
Improvements and enhancements
        - Upgraded to newer Django
          requires adaption of flowspy/settings.py regarding staticfiles handling
          (check 'git log -p flowspy/settings.py.dist')

        - Fixed unwanted possibility to mix IPv4 and IPv6 address prefixes in rules
        - Fixed broken check_sync for IPv6 rules
        - Support for implicit icmp6 in IPv6 rules 
        - Fixed improper UTF-8 encoding of Shibboleth attribute values
        - Rule id shown on rule detail pages

        - Installation support for DEBIAN/UBUNTU, including Systemd support
        - Cleanup of log file handling
        - Update of documentation
        - Enhancement of the Dockerfiles for testing and reference-installation
          - DEBIAN as base Docker image is now default: ./Dockerfile 
            (being linked to ./Dockerfile.fod.debian)
          - CENTOS7 keeps to be supported: ./Dockerfile.fod.centos.new

        - New experimental setting variable RULE_NAME_PREFIX:
          possibility to prefix every FlowSpec rule 
          injected via JUNOS_specific NETCONF on the router with a given String, 
          to distinguish such FlowSpec rules from 
          FlowSpec rules injected via NETCONF by other tools or means

=======
2022/08
Many improvements and enhancements
        - IPv6 rule support added
        - Improved IP range matching between rules and peer IP ranges

=======
1.7 RELEASE (in 2021)
Completed Reimplementation
        - Update to Python 3 and Django 2
        - Reimplementation of many parts necessary because of changed dependencies
        - REST API improved

=======
1.65 (in 2020)
Basic Reimplementation
        - Update to Python 3 and Django 2

=======
(in 2019)
Enhancements
        - Docker test installation container 
        - NETCONF simulator container for testing 

=======
1.5 RELEASE (in 2018)
Updates and Enhancements
        - Added SNMP statistics and statistic graphs
        - Easy specification of port ranges 
        - Improved date picker
        - eduGAIN errors made depending on the configuration
        - Many other enhancements

=======
1.2 RELEASE
Updates and enhancements. Check documentation for updating from previous versions
        - Code Cleanup
        - Branding
        - Rest Api
        - Easier Configuration

=======
1.1.1 RELEASE
Minor changes release

        - Change license to GPLv3

        - Minor documentation updates

===========
1.1.0 RELEASE
Updates and enhancements. Check documentation for updating from previous versions

        - Minor UI enhancements
        - Patch user model to include user peer in __unicode__
        - Include celery task exceptions in proxy
        - Cleanup js files
        - Resolve a major issue where the platform would start on 2nd refresh
        - Cleanup poller urls
        - Enhancements to json routes export
        - Include timeout handling in tasks
        - Fix issue with ports not updating
        - Add longerusername plugin to replace user monkey patching
        - Handle a bug in South that causes the Peer table to become unusable after adding autoincrement

===========
1.0.3 RELEASE
Minor fixes. Check documentation for peers table handling

        - Fix issue with altlogin redirection
        - Switch peers primary key to AutoField
        - Make peers tables management configurable

===========
1.0.2 RELEASE
Documentation enhancements

===========
1.0.1 RELEASE
Minor fixes

Fixes:
        -Overview login theme
        -Add missing urls

===========
1.0.0 RELEASE
Major UI redesign, Debian Wheezy version, Django 1.4

Improvements:
        -New UI based on Bootstrap3 theme
        -Minor fixes in long polling init
        -Debian Wheezy - Django 1.4 ready

===========
0.9.9 RELEASE
Major documentation improvements. Minor app enhancements. Minor bug fixes

Improvements:
        -Wrote documentation in rst format (http://flowspy.readthedocs.org/)
        -Update initial data with fragment types
        -Add current version in footer via context
        -Add beanstalk client, as installation from package is fuzzy
        -Comment the helptext line in patched user model (django 1.3 issues)

===========
0.9.8 RELEASE
Minor functionality improvements (check Requirements)

Requirements:
        -south migration to include database changes if you are at
        <=0.9.5

Improvements:
        -Added more port validation checks
        -Added more IP address validation checks

===========
0.9.7 RELEASE
Minor UI improvements (check Requirements)

Requirements:
        -south migration to include database changes if you are at
        <=0.9.5

UI Improvements:
        -Added badges in rule status

===========
0.9.6 RELEASE
New Feature and minor UI improvements (check Requirements)

Requirements:
        -south migration to include database changes

Features:
        -Added fragment type as an option in rule match statements

UI Improvements:
        -Changed wording;from 'Suspend' to 'Deactivate'
        -Increased the size of Console and Add Rule buttons. Made Add Rule button
        stand out with different color.

===========
0.9.5 RELEASE
Fixes

Fixes:
        -Fixed issue with page logo
        -Changed Shibboleth attributes from HTML to Shibboleth naming in error.html
        -Minor change in the user activation procedure. Activation mail goes only to admins not TechCs

===========
0.9.4 RELEASE
Minor fixes

Fixes:
        -Change the name of the released file (Makefile)
        -Added copyright info plus updated the README file
        -Added missing files in images

===========
0.9.3 RELEASE
Minor fix

Fixes:
        -Fixed the population of "Any" in source field

===========
0.9.2 RELEASE
Major enhancement and a minor fixes

Enhancements:
        -Added alternative view for helpdesk

Fixes:
        -Fixed the static url for tinymce in settings
        -Fixed an issue caused by multiple Shibboleth attributes

===========
0.9.1 RELEASE

inst/old/apache_shib/files.inst/srv/flowspy/flowspy/settings.py
# -*- coding: utf-8 -*- vim:fileencoding=utf-8:
# vim: tabstop=4:shiftwidth=4:softtabstop=4:expandtab
# Django settings for flowspy project.
# Copyright © 2011-2015 Greek Research and Technology Network (GRNET S.A.)
# Copyright © 2011-2014 Leonidas Poulopoulos (@leopoul)
# Copyright © 2014-2015 Stavros Kroustouris (@kroustou)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# _uuid_generate_random was deprecated and removed in newer python
import uuid
uuid._uuid_generate_random = None
try:
    from uuid import _uuid_generate_random
except ImportError:
    _uuid_generate_random = None

import os
import djcelery
djcelery.setup_loader()
from celery.schedules import crontab

DEBUG = True
TEMPLATE_DEBUG = DEBUG

ADMINS = (
    ('AdminName', 'admin@example.com'),
)
MANAGERS = ADMINS
here = lambda x: os.path.join(os.path.abspath(os.path.dirname(__file__)), x)
BASE_DIR = os.path.dirname(os.path.dirname(__file__))

# Hosts/domain names that are valid for this site; required if DEBUG is False
ALLOWED_HOSTS = ['*']
SITE_ID = 1

# Place a sequence of random chars here
SECRET_KEY = '@sa@5234#$%345345^@#$%*()123^@12!&!()$JMNDF#$@(@#8FRNJWX_'

# Set up database

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',  # Add 'postgresql_psycopg2', 'mysql', 'sqlite3' or 'oracle'.
        'NAME': 'example-data',
        'USER': '',
        'PASSWORD': '',
        'HOST': '',                      # Set to empty string for localhost.
        'PORT': '',                      # Set to empty string for default.
    }
}

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
#TIME_ZONE = 'Europe/Athens'
TIME_ZONE = 'UTC'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
_ = lambda s: s

LANGUAGES = (
    ('el', _('Greek')),
    ('en', _('English')),
)

LOCALE_PATHS = (
    os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'locale'),
)

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

PROJECT_PATH = os.path.abspath(os.path.dirname(__file__))
STATIC_ROOT = os.path.join(BASE_DIR)
STATICFILES_DIRS = ( os.path.join(BASE_DIR, 'static'), )
STATIC_URL = '/static/'
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

# Templates
TEMPLATE_DIRS = (
    os.path.join(BASE_DIR, 'templates'),
    os.path.join('/srv/venv/django/contrib/admin/templates/'),
)

TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
)

TEMPLATE_CONTEXT_PROCESSORS = (
    "django.contrib.auth.context_processors.auth",
    "django.core.context_processors.debug",
    "django.core.context_processors.i18n",
    "django.core.context_processors.media",
    "django.core.context_processors.static",
    "django.contrib.messages.context_processors.messages",
    "context.global_vars.settings_vars",
)

MIDDLEWARE_CLASSES = (
    'django.middleware.cache.UpdateCacheMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.cache.FetchFromCacheMiddleware',
    'django.contrib.flatpages.middleware.FlatpageFallbackMiddleware',
)

ROOT_URLCONF = 'flowspy.urls'
WSGI_APPLICATION = 'flowspy.wsgi.application'

AUTHENTICATION_BACKENDS = (
    'djangobackends.shibauthBackend.shibauthBackend',
    'django.contrib.auth.backends.ModelBackend',
)

INSTALLED_APPS = (
    'longerusername',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.flatpages',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'flowspec',
    'poller',
    'south',
    'django.contrib.admin',
    'djcelery',
    'peers',
    'registration',
    'accounts',
    'tinymce',
    'widget_tweaks',
    'rest_framework',
    'rest_framework.authtoken',
)

#GRAPHS_API_URL = 'graphs'
GRAPHS_API_URL = 'http://127.0.0.1:8080/api/routes/'

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'file': {
            'level': 'DEBUG',
            'class': 'logging.FileHandler',
            'filename': 'debug.log',
        },
    },
    'loggers': {
        'django': {
            'handlers': ['file'],
            'level': 'DEBUG',
            'propagate': True,
        },
    },
}

LOGIN_URL = '/welcome'

# CACHES = {
#     'default': {
#         'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
#         'LOCATION': '127.0.0.1:11211',
#     }
# }
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.dummy.DummyCache',
    }
}

AUTH_PROFILE_MODULE = 'accounts.UserProfile'

# Netconf Device credentials
# The following variables should contain the address of the device with
# flowspec, the username and password of the appliers account and the netconf
# port.
NETCONF_DEVICE = ""
NETCONF_USER = ""
NETCONF_PASS = ""
NETCONF_PORT = 830

# The filter applied in the device in order to find the flowspec routes
ROUTES_FILTER = "<configuration><routing-options><flow/></routing-options></configuration>"
ROUTE_FILTER = "<configuration><routing-options><flow><route><name>%s</name></route></flow></routing-options></configuration>"
COMMIT_CONFIRMED_TIMEOUT = "120"

# Change the following values only if you know what you are doing!!!
# To integrate FoD with tables (Peer, Networks, Contacts)
# from your CRM platform, set the following values to False and create the views that are
# exact matches of the tables in peers/models.py
PEER_MANAGED_TABLE = True
PEER_RANGE_MANAGED_TABLE = True
PEER_TECHC_MANAGED_TABLE = True

# Actually apply the rules
COMMIT = True

# Flowspy configuration
#EXPIRATION_DAYS_OFFSET = 7
EXPIRATION_DAYS_OFFSET = 30

USE_X_FORWARDED_HOST = True

# Beanstalk configuration (as a broker for celery)
BROKER_HOST = "localhost"
BROKER_PORT = 11300
POLLS_TUBE = 'polls'
BROKER_URL = "beanstalk://localhost:11300//"
BROKER_VHOST = "/"
CELERY_CONCURRENCY = 1
POLL_SESSION_UPDATE = 60.0

CELERYBEAT_SCHEDULE = {
    "every-day-sync": {
        "task": "flowspec.tasks.check_sync",
        "schedule": crontab(minute=01, hour=01),
        "args": (),
    },
    "notify-expired": {
        "task": "flowspec.tasks.notify_expired",
        "schedule": crontab(minute=01, hour=02),
        "args": (),
    },
    "snmp-stats-poll": {
        "task": "flowspec.tasks.poll_snmp_statistics",
        "schedule": crontab(minute="*/5"),
        "args": (),
    },
}

# List of modules to import when celery starts.
CELERY_IMPORTS = ("flowspec.tasks", )

# Notifications
DISABLE_EMAIL_NOTIFICATION = True
SERVER_EMAIL = "Example FoD Service <noreply@example.com>"
EMAIL_SUBJECT_PREFIX = "[FoD] "
EXPIRATION_NOTIFY_DAYS = 4
# max number of days into the future that is allowed to pick in rule expiration datepicker
MAX_RULE_EXPIRE_DAYS = 30
PREFIX_LENGTH = 29
POLL_SESSION_UPDATE = 60.0

# Shibboleth
#SHIB_AUTH_ENTITLEMENT = 'urn:mace:example.com:pki:user'
SHIB_AUTH_ENTITLEMENT = ''
SHIB_ADMIN_DOMAIN = 'example.com'
SHIB_LOGOUT_URL = 'https://example.com/Shibboleth.sso/Logout'

##############################################################################
# Shibboleth attribute map

# general info on these shibboleth attribute definitions:
# *_DISPLAY_NAME and *_ADDINFO should give edugain-global names and definitions and so match definitions in /etc/shibboleth/attribute-map.xml
# as values of the * variables (SHIB_ENTITLEMENT, SHIB_USERNAME, SHIB_MAIL, SHIB_FIRSTNAME, SHIB_LASTNAME) are in principle only host-config-local keys
#
# values of *_DISPLAY* are shown in edugain/login-related error messages for users; 
# instead, the values of * variables (e.g. SHIB_ENTITLEMENT, ...) are of no use for users and just confusing!
#
# specifically:
# *_DISPLAY_NAME should be a single string fully identifying the required attribute in an human-readable fashion
# *_DISPLAY_ADDINFO should be a string and should correspondingly contain descriptions for each array element of the value of the * variable

# e.g. for SHIB_ENTITLEMENT = "HTTP_ENTITLEMENT" => 
# 
# key in /etc/shibboleth/attribute-map.xml is "entitlement":
#    <Attribute name="urn:mace:dir:attribute-def:eduPersonEntitlement" id="entitlement"/>
#    <Attribute name="urn:oid:1.3.6.1.4.1.5923.1.1.1.7" id="entitlement"-->

#SHIB_ENTITLEMENT = ['HTTP_SHIB_EP_ENTITLEMENT'] # value in settings.py.dist
SHIB_ENTITLEMENT = ['HTTP_ENTITLEMENT']
SHIB_ENTITLEMENT_DISPLAY_NAME = "eduPersonEntitlement"
SHIB_ENTITLEMENT_DISPLAY_ADDINFO = "urn:oid:1.3.6.1.4.1.5923.1.1.1.7; the value of this attribute also has to include 'urn:mace:example.com:pki:user'"
#SHIB_ENTITLEMENT_DISPLAY_ADDINFO = "urn:oid:1.3.6.1.4.1.5923.1.1.1.7"

# (";"-separated) part of value of attribute with key SHIB_ENTITLEMENT needed in order to be considered a valid user:
# if SHIB_AUTH_ENTITLEMENT is empty no constraint on SHIB_ENTITLEMENT attribute value is enforced (can even be missing)
#SHIB_AUTH_ENTITLEMENT = 'urn:mace:example.com:pki:user'
SHIB_AUTH_ENTITLEMENT = '' # is also in settings_local.py so override it there

#SHIB_USERNAME = ['HTTP_EPPN'] # in settings.py.dist # originally used for SHIB_USERNAME
#SHIB_USERNAME_DISPLAY_NAME = "eduPersonPrincipalName"
#SHIB_USERNAME_DISPLAY_ADDINFO = "urn:mace:dir:attribute-def:eduPersonPrincipalName or urn:oid:1.3.6.1.4.1.5923.1.1.1.6"
#SHIB_USERNAME = ['HTTP_PERSISTENT_ID']
SHIB_USERNAME = ['HTTP_PERSISTENT_ID', 'HTTP_PRINCIPALNAME']
#SHIB_USERNAME = ['mail', 'HTTP_MAIL', 'HTTP_SHIB_INETORGPERSON_MAIL']
SHIB_USERNAME_DISPLAY_NAME = "PERSISTENT_ID"
SHIB_USERNAME_DISPLAY_ADDINFO = "urn:oid:1.3.6.1.4.1.5923.1.1.1.10"

#SHIB_SLUGIFY_USERNAME = False # value in settings.py.dist
SHIB_SLUGIFY_USERNAME = True

SHIB_MAIL = ['mail', 'HTTP_MAIL', 'HTTP_SHIB_INETORGPERSON_MAIL']
SHIB_MAIL_DISPLAY_NAME = "MAIL"
SHIB_MAIL_DISPLAY_ADDINFO = "urn:mace:dir:attribute-def:mail or urn:oid:0.9.2342.19200300.100.1.3 or SHIB_INETORGPERSON_MAIL"

# both optional:
SHIB_FIRSTNAME = ['HTTP_SHIB_INETORGPERSON_GIVENNAME'] # value in settings.py.dist
#SHIB_FIRSTNAME = ['HTTP_GIVENNAME']
SHIB_FIRSTNAME_DISPLAY_NAME = "GIVENNAME"
SHIB_FIRSTNAME_DISPLAY_ADDINFO = "urn:mace:dir:attribute-def:givenName or urn:oid:2.5.4.42"

SHIB_LASTNAME = ['HTTP_SHIB_PERSON_SURNAME'] # value in settings.py.dist
#SHIB_LASTNAME = ['HTTP_SURNAME']
SHIB_LASTNAME_DISPLAY_NAME = "SURNAME"
SHIB_LASTNAME_DISPLAY_ADDINFO = "urn:mace:dir:attribute-def:sn or urn:oid:2.5.4.4"

##############################################################################

# BCC mail addresses
NOTIFY_ADMIN_MAILS = ["admin@example.com"]

# Then actions in the ui (dropdown)
UI_USER_THEN_ACTIONS = ['discard', 'rate-limit']
UI_USER_PROTOCOLS = ['icmp', 'tcp', 'udp']
ACCOUNT_ACTIVATION_DAYS = 7

# Define subnets that should not have any rules applied whatsoever
PROTECTED_SUBNETS = ['10.10.0.0/16']

# Add two whois servers in order to be able to get all the subnets for an AS.
PRIMARY_WHOIS = 'whois.example.com'
ALTERNATE_WHOIS = 'whois.example.net'

TINYMCE_JS_URL = STATIC_URL + 'js/tinymce/tiny_mce.js'

TINYMCE_DEFAULT_CONFIG = {
    'extended_valid_elements': 'iframe[src|width|height|name|align]',
    'plugins': "table,spellchecker,paste,searchreplace",
    'theme': "advanced",
}

import _version
SW_VERSION = _version.VERSION

LOG_FILE_LOCATION = "/var/log/fod"


BRANDING = {
    'name': 'Example',
    'url': 'https://example.com',
    'footer_iframe': 'https://example.com/iframe',
    'facebook': '//facebook.com/',
    'twitter': '//twitter.com/',
    'phone': '800-example-com',
    'email': 'helpdesk@example.com',
    'logo': 'fodlogo2.png',
    'favicon': 'favicon.ico',
}


# Django Rest Framework configuration.
# You should leave this intact.
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.TokenAuthentication',
    ),
    'DEFAULT_RENDERER_CLASSES': (
        'rest_framework.renderers.JSONRenderer',
    ),
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.DjangoModelPermissions',
        'rest_framework.permissions.IsAuthenticated'
    ]
}

# Limit of ports in 'ports' / 'SrcPorts' / 'DstPorts' of a rule:
#PORTRANGE_LIMIT = 100
PORTRANGE_LIMIT = 65535

# Statistics polled via SNMP:
# Default community string
SNMP_COMMUNITY = "abcd"

# list of IP addresses, each IP is a dict with "ip", "port" (optional, default
# is 161), "community" (optional, default is SNMP_COMMUNITY) keys
SNMP_IP = [
    {"ip": "192.168.0.1", "port": 1000},
    {"ip": "192.168.0.2", "port": 1001, "community": "abcdef"},
    {"ip": "192.168.0.3", "port": 1002},
    {"ip": "192.168.0.4", "port": 1002}
]

# or simpler way of IP list:
# SNMP_IP = ["10.0.0.1", "10.0.0.2"]

# OID of bytes counter (currently unused)
SNMP_CNTBYTES =     "1.3.6.1.4.1.2636.3.5.2.1.5"
# OID of packet counter
SNMP_CNTPACKETS =   "1.3.6.1.4.1.2636.3.5.2.1.4"

# get only statistics of specified tables
SNMP_RULESFILTER = ["__flowspec_default_inet__", "__flowspec_IAS_inet__"]
# load new data into cache if it is older that a specified number of seconds
SNMP_POLL_INTERVAL = 8 #seconds
# cache file for data
SNMP_TEMP_FILE = "/tmp/snmp_temp_data"
SNMP_POLL_LOCK = "/var/run/fod/snmppoll.lock"

# Number of historical values to store for a route.
# Polling interval must be set for "snmp-stats-poll" celery task in CELERYBEAT_SCHEDULE.
# By default, it is 5 min interval, so SNMP_MAX_SAMPLECOUNT=12 means we have about
# one hour history.
SNMP_MAX_SAMPLECOUNT = 12

# Age of inactive routes that can be already removed (in seconds)
SNMP_REMOVE_RULES_AFTER = 3600

##############################################################################
##############################################################################

# REST API config (v1.6 only)

DISABLE_RULE_OVERLAP_CHECK = False

ALLOW_DELETE_FULL_FOR_NONADMIN = False

MAIL_NOTIFICATION_TO_ALL_MATCHING_PEERS = True

# statistics calc

STATISTICS_PER_MATCHACTION_ADD_FINAL_ZERO = False # not necessary if STATISTICS_PER_RULE==True
STATISTICS_PER_RULE = True

STATISTICS_PER_RULE__ADD_INITIAL_ZERO = True

##############################################################################
##############################################################################

#MYSETTING1="default"
#from settings_local import *
from flowspy.settings_local import *

#print "MYSETTING1="+MYSETTING1
#print("MYSETTING1="+MYSETTING1, file=sys.stderr)
#print "debug settings.NOTIFY_ADMIN_MAILS="+str(NOTIFY_ADMIN_MAILS)

##############################################################################
##############################################################################

